tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,577 @@
+      String error = doPrewriteSecondaryKeySingleBatch(backOffer, primaryKey, batchKeys, mutations);
+      if (error != null) {
+        return error;
+      }
+    }
+    return null;
+  }
+
+  private String doPrewriteSecondaryKeySingleBatch(
+      BackOffer backOffer,
+      byte[] primaryKey,
+      BatchKeys batchKeys,
+      Map<ByteWrapper, Kvrpcpb.Mutation> mutations) {
+    List<byte[]> keyList = batchKeys.getKeys();
+    int batchSize = keyList.size();
+    byte[][] keys = new byte[batchSize][];
+    int index = 0;
+    List<Kvrpcpb.Mutation> mutationList = new ArrayList<>(batchSize);
+    for (byte[] key : keyList) {
+      mutationList.add(mutations.get(new ByteWrapper(key)));
+      keys[index++] = key;
+    }
+    // send rpc request to tikv server
+    long regionId = batchKeys.getRegioId();
+    int txnSize = batchKeys.getKeys().size();
+    long lockTTL = getTxnLockTTL(this.startTs, txnSize);
+    ClientRPCResult prewriteResult =
+        this.kvClient.prewrite(
+            backOffer, mutationList, primaryKey, lockTTL, this.startTs, regionId);
+    if (!prewriteResult.isSuccess() && !prewriteResult.isRetry()) {
+      return prewriteResult.getError();
+    }
+    if (!prewriteResult.isSuccess() && prewriteResult.isRetry()) {
+      try {
+        backOffer.doBackOff(
+            BackOffFunction.BackOffFuncType.BoRegionMiss,
+            new GrpcException(
+                String.format(
+                    "Txn prewrite secondary key SingleBatch failed, regionId=%s, detail=%s",
+                    batchKeys.getRegioId(), prewriteResult.getError())));
+        // re-split keys and commit again.
+        int size = batchKeys.getKeys().size();
+        byte[][] keyBytes = new byte[size][];
+        byte[][] valueBytes = new byte[size][];
+        int i = 0;
+        for (byte[] k : batchKeys.getKeys()) {
+          keyBytes[i] = k;
+          valueBytes[i] = mutations.get(new String(k)).getValue().toByteArray();
+          i++;
+        }
+        return doPrewriteSecondaryKeys(backOffer, primaryKey, keyBytes, valueBytes, size);
+      } catch (GrpcException e) {
+        String error =
+            String.format(
+                "Txn prewrite secondary key SingleBatch error, re-split commit failed, regionId=%s, detail=%s",
+                batchKeys.getRegioId(), e.getMessage());
+        LOG.error(error);
+        return error;
+      }
+    }
+    // success return
+    LOG.info("prewrite " + batchKeys.getKeys().size() + "rows successfully");
+    return null;
+  }
+
+  private void appendBatchBySize(
+      List<BatchKeys> batchKeyList,
+      Long regionId,
+      List<byte[]> keys,
+      boolean sizeKeyValue,
+      int limit,
+      Map<ByteWrapper, Kvrpcpb.Mutation> mutations) {
+    int start;
+    int end;
+    int len = keys.size();
+    for (start = 0; start < len; start = end) {
+      int size = 0;
+      for (end = start; end < len && size < limit; end++) {
+        if (sizeKeyValue) {
```

Comment:
where will the sizeKeyValue be used?

Reply:
prewrite: sizeKeyValue = true
commit: sizeKeyValue = false